### PR TITLE
feat!: Introduce Jacobian engine, consistent Jacobian naming

### PR DESCRIPTION
**Pull request by asalzburger**
This PR is a cherry-picked version of #751 

This PR will (eventually) establish the covariance transport for all parameterisations, it also introduces a consistent naming (hence the many changed files).

BREAKING CHANGE:

Local is only referred to for a two-dimensional local representation

Global is only referred to for three-dimensional global representation

bound refers to: local + angular + q/p + time

free refers to: global + time + direction + time

Hence, a transformation of a Vector2 to Vector3 is localToGlobal, but a Jacobian from a bound/curvilinear representation to a free one is boundToFree and vice versa (was also called localToGlobal and globalToLocal).

This PR does not change the covariance transport yet, it introduces the right classes to do so.

I added a benchmark to see if either variant or optional shows up, looks all good, the time is totally dominated by the matrix multiplication (as it should be).

### CODE REVIEW

**Comment by asalzburger** [comment]
Hi Fabian, this is a squashed and cherry-picked version of the original that addresses basically all comments.

I am puzzled by the reporting of code coverage, because it basically tests every new function ...

**Review by FabianKlimpel — COMMENTED** [review]
Found some typos but in general it's ready to go in

File: Core/include/Acts/Propagator/detail/JacobianEngine.hpp [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+/// @brief This function calculates the full jacobian from a given
+/// free parameterisation to a new curvilinear bound
+/// parameterisation.
+///
+/// @note Modifications of the jacobian related to the
+/// projection onto the target surface is considered. Since a variation of
+/// the start parameters within a given uncertainty would lead to a variation of
+/// the end parameters, these need to be propagated onto the target surface.
+/// This an approximated approach to treat the (assumed) small change.
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Core/include/Acts/Propagator/detail/JacobianEngine.hpp:132]: Fixed with all occurrences.

File: Core/include/Acts/Utilities/Helpers.hpp [hunk 2]
```diff
@@ -121,6 +122,26 @@ double theta(const Eigen::MatrixBase<Derived>& v) noexcept {
   return std::atan2(std::sqrt(v[0] * v[0] + v[1] * v[1]), v[2]);
 }
 
+/// @brief Fast evaluation of trigonomic functions.
+///
+/// @param direction for this evaluatoin
+///
+/// @return cos(phi), sin(phi), cos(theta), sin(theta), 1/sin(theta)
+static inline const std::array<ActsScalar, 5> evaluateTrigonimics(
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Core/include/Acts/Utilities/Helpers.hpp:130]: Fixed wherever this is used.

File: Tests/UnitTests/Core/Propagator/JacobianEngineTests.cpp [hunk 3]
```diff
@@ -0,0 +1,318 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Tests/UnitTests/Core/Propagator/JacobianEngineTests.cpp:3]: Also fixed.

**Review by asalzburger — COMMENTED** [review]
@FabianKlimpel  I think that should address all comments

**Comment by asalzburger** [comment]
Ready to review, could not find out what's wrong with the codev.

**Review by FabianKlimpel — APPROVED** [review]
Looks good to me.
I didn't check the math again and trust you that it's still correct.

**Pull request merged**